src/unix.rs

```diff
@@ -7,27 +7,204 @@
+            sender: sender,
+            receiver: receiver,
+            signals: Default::default(),
+            drivers: Mutex::new(HashSet::new()),
+        }
+    }
+}
+
+lazy_static! {
+    // TODO: Get rid of lazy_static once the prototype is done ‒ get rid of the dependency as well
+    // as the possible lock in there, which *might* be problematic in signals
+    static ref GLOBALS: Globals = Globals::new();
+}
+
+// Flag the relevant signal and wake up through a self-pipe
+extern "C" fn pipe_wakeup(signal: c_int) {
+    let index = signal as usize;
+    // TODO: Handle the old signal handler
+    // It might be good enough to use some lesser ordering than this, but how to prove it?
+    GLOBALS.pending[index].store(true, Ordering::SeqCst);
+    // Send a wakeup, ignore any errors (anything reasonably possible is full pipe and then it will
+    // wake up anyway).
+    let _ = send(GLOBALS.sender, &[0u8], MSG_DONTWAIT);
+}
+
+// Make sure we listen to the given signal and provide the recipient end of the self-pipe
+fn signal_enable(signal: c_int) {
+    let index = signal as usize;
+    let mut siginfo = GLOBALS.signals[index].lock().unwrap();
+    if !siginfo.initialized {
+        let action = SigAction::new(SigHandler::Handler(pipe_wakeup), SA_NOCLDSTOP | SA_RESTART, SigSet::empty());
+        unsafe { sigaction(NixSignal::from_c_int(signal).unwrap(), &action).unwrap() };
+        // TODO: Handle the old signal handler
+        siginfo.initialized = true;
+    }
+}
+
+struct EventedReceiver;
+
+impl Evented for EventedReceiver {
+    fn register(&self, poll: &MioPoll, token: Token, events: Ready, opts: PollOpt) -> io::Result<()> {
+        EventedFd(&GLOBALS.receiver).register(poll, token, events, opts)
+    }
+    fn reregister(&self, poll: &MioPoll, token: Token, events: Ready, opts: PollOpt) -> io::Result<()> {
+        EventedFd(&GLOBALS.receiver).reregister(poll, token, events, opts)
+    }
+    fn deregister(&self, poll: &MioPoll) -> io::Result<()> {
+        EventedFd(&GLOBALS.receiver).deregister(poll)
+    }
+}
+
+// There'll be stuff inside
+struct Driver {
+    id: CoreId,
+    wakeup: PollEvented<EventedReceiver>,
+}
+
+impl Future for Driver {
+    type Item = ();
+    type Error = ();
+
+    fn poll(&mut self) -> Poll<(), ()> {
+        // Drain the data from the pipe and maintain interest in getting more
+        let any_wakeup = self.drain();
+        if any_wakeup {
+            self.broadcast();
+        }
+        // This task just lives until the end of the event loop
+        Ok(Async::NotReady)
+    }
+}
 
+impl Drop for Driver {
+    fn drop(&mut self) {
+        let mut drivers = GLOBALS.drivers.lock().unwrap();
+        drivers.remove(&self.id);
+    }
+}
+
+impl Driver {
+    fn new(handle: &Handle) -> Self {
+        Driver {
+            id: handle.id(),
+            // TODO: Any chance of errors here?
+            wakeup: PollEvented::new(EventedReceiver, handle).unwrap(),
```

Comment:
We may as well propagate this upwards, but I don't think there's much of a risk of an error